cpp/src/arrow/compute/exec/options.h

```diff
@@ -397,23 +397,33 @@ class ARROW_EXPORT HashJoinNodeOptions : public ExecNodeOptions {
 /// This node will output one row for each row in the left table.
 class ARROW_EXPORT AsofJoinNodeOptions : public ExecNodeOptions {
  public:
-  AsofJoinNodeOptions(FieldRef on_key, FieldRef by_key, int64_t tolerance)
-      : on_key(std::move(on_key)), by_key(std::move(by_key)), tolerance(tolerance) {}
+  AsofJoinNodeOptions(FieldRef on_key, const FieldRef& by_key, int64_t tolerance)
+      : on_key(std::move(on_key)), by_key(), tolerance(tolerance) {
+    this->by_key.push_back(std::move(by_key));
+  }
+
+  AsofJoinNodeOptions(FieldRef on_key, std::vector<FieldRef> by_key, int64_t tolerance)
+      : on_key(std::move(on_key)), by_key(by_key), tolerance(tolerance) {}
+
+  // resolves ambiguity between previous constructors when initializer list is given
+  AsofJoinNodeOptions(FieldRef on_key, std::initializer_list<FieldRef> by_key,
+                      int64_t tolerance)
+      : on_key(std::move(on_key)), by_key(by_key), tolerance(tolerance) {}
 
   /// \brief "on" key for the join. Each
   ///
-  /// All inputs tables must be sorted by the "on" key. Inexact
-  /// match is used on the "on" key. i.e., a row is considiered match iff
+  /// All inputs tables must be sorted by the "on" key. Must be a single field of a common
+  /// type. Inexact match is used on the "on" key. i.e., a row is considered match iff
   /// left_on - tolerance <= right_on <= left_on.
-  /// Currently, "on" key must be an int64 field
+  /// Currently, the "on" key must be of an integer or timestamp type
   FieldRef on_key;
   /// \brief "by" key for the join.
   ///
   /// All input tables must have the "by" key.  Exact equality
   /// is used for the "by" key.
-  /// Currently, the "by" key must be an int32 field
-  FieldRef by_key;
-  /// Tolerance for inexact "on" key matching
+  /// Currently, the "by" key must be of an integer or timestamp type
+  std::vector<FieldRef> by_key;
+  /// Tolerance for inexact "on" key matching.  Must be non-negative.
   int64_t tolerance;
```

Comment:
What does tolerance mean here if the "on" key is timestamp? Is it interpreted as "number of nano seconds"?

Reply:
Or number of unit of the timestamp column?

Reply:
It's interpreted using the [time unit of the timestamp](https://github.com/apache/arrow/pull/13880/files#diff-5789a42aebc3bd0f5a6db687c78ab0c3eef3e316982f99fe9e0ea21fabed354cR1009-R1012).